mgenware/string-to-template-literal
# simplify package

this is a **very radical pr**. i want to emphasize you **do not have to merge this**. ❤️

this pr would:
- use `node --test` instead of mocha
- use plain js instead of typescript. **why?** because this function is **5 lines of code** and doesn't need a build step. it's small enough that the tradeoff of using tsc/whatever is not worth it imo
- combine tests into single file
- add npm-publish.yml that uses secrets.NPM_TOKEN to publish on each new release -- this is useful to avoid the "ahhh ill release it tomorrow" but **is not needed at all with a package of this scale**. you can delete this if you want.
- bumps to v3
- use nodejs 18, 20, 21 instead of 14 and 16 for the test.yml ci -- frankly you might not even need a test.yml workflow either lol since this is so tiny
- remove eslint; its not worth it at this scale
- use jsdoc instead of .d.ts for types -- still get type hints and doc comments in ide like vscode
- remove redundant gitattributes; the defaults are OK at this small scale
- use ZERO dependencies
- add note in readme about copy-pasting (since its 5 line function)
- add use cases where you might want to use this in Usage section of readme. points out how compares vs JSON.stringify()
- add header image/example to readme
- change license to 0bsd -- this way if a user does the copy-paste install method they dont need to include the mit license text (as required by the mit license) since its the 0bsd license which has no such "must retain license notice" clause in it
- **uses regex instead of iterative string building method**. this is more concise, faster, and understandable. there's a comment to explain the special lookahead/lookbehind for `${` and `</` cases.
- adds a case to **escape `</script>`-like strings so that this can be injected safely into `<script>` tags** when building an html page.
- more gracefully handles non-string inputs by coercing to string -- open question: how should `stringToTemplateLiteral()` be handled? `` `undefined` ``? ``` `` ```? right now it defaults to the empty string ``` `` ``` but maybe it should be `` `undefined` `` like most webapis do with required string args?
- since 0 deps this would stop the dependabot stuff like #10 #11 #12 #13

closes #15 

the code:
```js
/** @param {string} x */
export default function stringToTemplateLiteral(x = "") {
  // This coerces 'x' to a string.
  x = `${x}`;

  // \$(?={): Matches a dollar sign followed by an open curly brace (${).
  //    It's looking for the start of a template expression (e.g., ${...})
  //    but with an escaped dollar sign.
  //
  // (?<=<)\/: Uses lookbehind and lookahead assertions. It matches a
  //    forward slash (/) that is preceded by a less-than sign (<) such
  //    as '</script>'. It doesn't include the '<' in the match text.
  const escaped = x.replace(/\\|`|\$(?={)|(?<=<)\//g, (y) => "\\" + y);

  return `\`${escaped}\``;
}
```

feel free to edit this pr. the "allow edits by maintainers" is checked 👍


-----
<a href="https://stackblitz.com/~/github/jcbhmr/string-to-template-literal/tree/jcbhmr%2Fpatch-51470"><img src="https://developer.stackblitz.com/img/review_pr_small.svg" alt="Review PR in StackBlitz Codeflow" align="left" width="103" height="20"></a> _Submitted with [StackBlitz Codeflow](https://stackblitz.com/~/github/jcbhmr/string-to-template-literal/tree/jcbhmr%2Fpatch-51470)._
